payments/payments.go

```diff
@@ -0,0 +1,41 @@
+package payments
+
+import (
+	"context"
+	"net/http"
+)
+
+const (
+	StripeProvider = "stripe"
+	PayPalProvider = "paypal"
+)
+
+// Provider represents a payment provider that can optionally charge, refund,
+// preauthorize payments.
+type Provider interface {
```

Comment:
I was hoping that having separate interfaces meant the provider would only implement the ones they supported and a type assertion would allow an easy check for support. I wasn't able to find a way to implement that cleanly.

One catch is the request parameters requirements vary based on the action you are taking (charge, refund, or preauthorize). For example, Stripe needs a `stripe_token`, but only for charging. Checking for the required params upon creation also allowed fast failure in the API methods instead of waiting until the action was ready to be taken.